File: core/src/main/java/nl/sjtek/control/core/settings/SettingsManager.java
package nl.sjtek.control.core.settings;

import com.google.gson.Gson;
import io.habets.javautils.Log;
import nl.sjtek.control.core.utils.DummyData;
import nl.sjtek.control.core.utils.FileUtils;
import nl.sjtek.control.data.settings.*;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Map;

/**
 * Created by wouter on 22-11-15.
 */
public class SettingsManager {

    private static final String DEFAULT_PATH = "/var/sjtekcontrol/config.json";
    private static final String DEBUG = SettingsManager.class.getSimpleName();
    private static SettingsManager instance = DummyData.getSettingsManager();

    private final LoggingSettings logging;
    private final MusicSettings music;
    private final TVSettings tv;
    private final QuotesSettings quotes;
    private final LastFMSettings lastFM;
    private final Map<String, User> users;
    private final ScreenSettings screen;
    private final WeatherSettings weather;

    public SettingsManager(MusicSettings music,
                           TVSettings tv,
                           QuotesSettings quotes,
                           LastFMSettings lastFM,
                           Map<String, User> users,
                           ScreenSettings screen,
                           WeatherSettings weatherSettings,
                           LoggingSettings logging) {
        this.music = music;
        this.tv = tv;
        this.quotes = quotes;
        this.lastFM = lastFM;
        this.users = users;
        this.screen = screen;
        this.weather = weatherSettings;
        this.logging = logging;
    }

    public static SettingsManager getInstance() {
        return instance;
    }

    public String dump() {
        return new Gson().toJson(this, this.getClass());
    }

    public void reload() {
        reload(DEFAULT_PATH);
    }

    public void reload(String path) {
        Log.i(SettingsManager.class.getSimpleName(), "");
        SettingsManager newSettingsManager;
        try {
            String jsonString = FileUtils.readFile(path);
            if (!jsonString.isEmpty()) {
                newSettingsManager = new Gson().fromJson(jsonString, this.getClass());
                Log.i(DEBUG, "Settings reloaded (" + path + ")");
            } else {
                throw new IOException("Data empty");
            }
        } catch (FileNotFoundException e) {
            Log.e(DEBUG, "Could not load settings", e);
            try {
                FileUtils.writeFile(path, dump());
            } catch (IOException ignored) {
            }
            newSettingsManager = DummyData.getSettingsManager();
        } catch (IOException e) {
            Log.e(DEBUG, "Reload error", e);
            newSettingsManager = DummyData.getSettingsManager();
        }

        instance = newSettingsManager;
        Log.setListener(new Log.PrintListener(instance.getLoggingSettings().isPrintStackTrace()));
        Log.setLevel(Log.Level.valueOf(instance.getLoggingSettings().getLevel()));
        Log.i(DEBUG, "Log level " + instance.getLoggingSettings().getLevel() + " PrintStackTrace " + instance.getLoggingSettings().isPrintStackTrace());
    }

    public MusicSettings getMusic() {
        return music;
    }

    public TVSettings getTv() {
        return tv;
    }

    public QuotesSettings getQuotes() {
        return quotes;
    }

    public Map<String, User> getUsers() {
        return users;
    }

    public User getUser(String name) {
        return users.get(name.toLowerCase());
    }

    public User getDefaultUser() {
        return getUser("default");
    }

    public LastFMSettings getLastFM() {
        return lastFM;
    }

    public ScreenSettings getScreen() {
        return screen;
    }

    public WeatherSettings getWeather() {
        return weather;
    }

    public LoggingSettings getLoggingSettings() {
        return logging;
    }

    @Override
    public String toString() {
        return new Gson().toJson(this);
    }

}

File: core/src/main/kotlin/nl/sjtek/control/core/settings/SettingsManager.kt
package nl.sjtek.control.core.settings

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import nl.sjtek.control.data.parsers.LampHolder
import nl.sjtek.control.data.parsers.QuotesHolder
import nl.sjtek.control.data.parsers.UserHolder
import nl.sjtek.control.data.static.Lamp
import nl.sjtek.control.data.static.User
import org.slf4j.LoggerFactory
import java.io.File

object SettingsManager {
    private const val PATH_CONFIG = "/var/sjtekcontrol/config.json"
    private const val PATH_USERS = "/var/sjtekcontrol/users.json"
    private const val PATH_QUOTES = "/var/sjtekcontrol/quotes.json"
    private const val PATH_LAMPS = "/var/sjtekcontrol/lamps.json"
    private val logger = LoggerFactory.getLogger(javaClass)

    val settings: Settings
    val users: List<User>
    val quotes: List<String>
    val lamps: Map<Int, Lamp>

    val jsonUsers: String
        get() = GsonBuilder().setPrettyPrinting().create().toJson(users)
    val jsonQuotes: String
        get() = GsonBuilder().setPrettyPrinting().create().toJson(quotes)
    val jsonLamps: String
        get() = GsonBuilder().setPrettyPrinting().create().toJson(lamps)

    init {
        settings = load(PATH_CONFIG, Settings::class.java) ?: Settings()
        users = load(PATH_USERS, UserHolder::class.java)?.users ?: listOf()
        logger.info("Loaded ${users.size} users")
        quotes = load(PATH_QUOTES, QuotesHolder::class.java)?.quotes ?: listOf()
        logger.info("Loaded ${quotes.size} quotes")
        lamps = load(PATH_LAMPS, LampHolder::class.java)?.lamps ?: mapOf()
        logger.info("Loaded ${lamps.size} lamps")
    }

    fun getUser(name: String?): User? {
        if (name == null) return null
        return users.find { it.username == name }
    }

    fun getUser(request: spark.Request): User? {
        val userName = request.queryParams("user")
        return getUser(userName)
    }

    private fun <T> load(path: String, clazz: Class<T>): T? {
        val jsonString = try {
            File(path).readText()
        } catch (e: Exception) {
            logger.error("Failed loading $path", e)
            return null
        }

        return try {
            Gson().fromJson<T>(jsonString, clazz)
        } catch (e: Exception) {
            logger.error("Failed parsing $path", e)
            null
        }
    }
}